### PR TITLE
Multivalue Feature - implement multi-value Brute Force Algorithm

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR summarizes the effort for implementing a multi-value BF algorithm.
These PRs do not include `RANGE` multi-value support.

**Included Commits**
1. #204
2. #206
3. #219
4. #226

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: tests/unit/test_bruteforce_multi.cpp [hunk 1]
```diff
@@ -0,0 +1,1492 @@
...
+    ASSERT_EQ(VecSimIndex_IndexSize(index), 0);
+    ASSERT_EQ((reinterpret_cast<BruteForceIndex_Multi<float, float> *>(index))->indexLabelCount(),
+              0);
+
+    VecSimIndex_Free(index);
+}
+
+/**** resizing cases ****/
+
+TEST_F(BruteForceMultiTest, resizeNAlignIndex) {
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce_multi.cpp:56]:
```suggestion
TEST_F(BruteForceMultiTest, resizeAndAlignIndex) {
```

File: tests/unit/test_bruteforce_multi.cpp [hunk 2]
```diff
@@ -0,0 +1,1492 @@
...
+    dist = VecSimIndex_GetDistanceFrom(index[VecSimMetric_L2], 46, query);
+    ASSERT_TRUE(std::isnan(dist));
+
+    // Clean-up.
+    for (size_t i = 0; i < numIndex; i++) {
+        VecSimIndex_Free(index[i]);
+    }
+}
+
+// TEST_F(BruteForceMultiTest, testCosine) {
```
Comment (DvirDukhan) [inline tests/unit/test_bruteforce_multi.cpp:1165]: what about this one?

File: src/VecSim/utils/updatable_heap.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+bool updatable_max_heap<Priority, Value>::empty() const {
+    return valueToPriority.empty();
+}
+
+template <typename Priority, typename Value>
+const std::pair<Priority, Value> updatable_max_heap<Priority, Value>::top() const {
+    // The `.begin()` of "priorityToValue" is the max priority element.
+    auto x = priorityToValue.begin();
+    return *x;
+}
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/updatable_heap.h:65]: I can add the `.end()` function to the class. I think in the empty case it should return the same value (in the standard class)

File: src/VecSim/utils/updatable_heap.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                       VecsimSTLAllocator<std::pair<const Value, Priority>>>
+        valueToPriority;
+
+public:
+    updatable_max_heap(const std::shared_ptr<VecSimAllocator> &alloc);
+    ~updatable_max_heap();
+
+    inline void emplace(Priority p, Value v) override;
+    inline bool empty() const override;
+    inline void pop() override;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/utils/updatable_heap.h:36]: The `.pop()` function of the standard priority queue is void, and the top value should be obtain by calling `.top()`

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 5]
```diff
@@ -185,11 +190,11 @@ int BruteForceIndex<DataType, DistType>::removeVector(idType id_to_delete) {
...
         this->vectorBlocks.pop_back();
 
-        // Resize and align the id2labelmapping.
+        // Resize and align the id2labelMapping.
         size_t id2label_size = idToLabelMapping.size();
-        // If the new size is smaller by at least one block comparing to the id2labelmapping
+        // If the new size is smaller by at least one block comparing to the id2labelMapping
         // align to be a multiplication of blocksize  and resize by one block.
-        if (count + this->blockSize <= id2label_size) {
+        if (this->count + this->blockSize <= id2label_size) {
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/algorithms/brute_force/brute_force.h:197]: Agree

File: tests/unit/test_bruteforce_multi.cpp (hunk 1 above)
Comment (GuyAv46) [reply tests/unit/test_bruteforce_multi.cpp:56]: Agree

File: tests/unit/test_bruteforce_multi.cpp (hunk 2 above)
Comment (GuyAv46) [reply tests/unit/test_bruteforce_multi.cpp:1165]: I just noticed that myself… I’ll add it in

**Comment by GuyAv46** [comment]
I’m going to squash the review changes into the last commit so I can merge to master without squashing. Are you ok with that @DvirDukhan?

**Comment by DvirDukhan** [comment]
Prefer not. It is hard to track changes post review.

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**